test/pinverse.cpp

```diff
@@ -0,0 +1,327 @@
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <gtest/gtest.h>
+#include <arrayfire.h>
+#include <af/dim4.hpp>
+#include <af/defines.h>
+#include <af/traits.hpp>
+#include <iostream>
+#include <complex>
+#include <testHelpers.hpp>
+#include <limits>
+
+using af::array;
+using af::cdouble;
+using af::cfloat;
+using af::constant;
+using af::dim4;
+using af::dtype;
+using af::dtype_traits;
+using af::exception;
+using af::identity;
+using af::matmul;
+using af::max;
+using af::pinverse;
+using af::randu;
+using af::span;
+using std::abs;
+using std::string;
+using std::vector;
+
+template<typename T>
+array makeComplex(dim4 dims, const vector<T>& real, const vector<T>& imag) {
+    array realArr(dims, &real.front());
+    array imagArr(dims, &imag.front());
+    return af::complex(realArr, imagArr);
+}
+
+template<typename T>
+array readTestInput(string testFilePath) {
+    typedef typename dtype_traits<T>::base_type InBaseType;
+    dtype outAfType = (dtype) dtype_traits<T>::af_type;
+
+    vector<dim4> dimsVec;
+    vector<vector<InBaseType> > inVec;
+    vector<vector<InBaseType> > goldVec;
+    readTestsFromFile<InBaseType, InBaseType>(testFilePath, dimsVec, inVec, goldVec);
+    dim4 inDims = dimsVec[0];
+
+    if (outAfType == c32 || outAfType == c64) {
+        return makeComplex(inDims, inVec[1], inVec[2]);
+    }
+    else {
+        return array(inDims, &inVec[0].front());
+    }
+}
+
+template<typename T>
+array readTestGold(string testFilePath) {
+    typedef typename dtype_traits<T>::base_type InBaseType;
+    dtype outAfType = (dtype) dtype_traits<T>::af_type;
+
+    vector<dim4> dimsVec;
+    vector<vector<InBaseType> > inVec;
+    vector<vector<InBaseType> > goldVec;
+    readTestsFromFile<InBaseType, InBaseType>(testFilePath, dimsVec, inVec, goldVec);
+    dim4 goldDims(dimsVec[0][1], dimsVec[0][0]);
+
+    if (outAfType == c32 || outAfType == c64) {
+        return makeComplex(goldDims, goldVec[1], goldVec[2]);
+    }
+    else {
+        return array(goldDims, &goldVec[0].front());
+    }
+}
+
+template<typename T>
+class Pinverse : public ::testing::Test
+{
+
+};
+
+// Epsilons taken from test/inverse.cpp
+template<typename T>
+double eps();
+
+template<>
+double eps<float>() {
+    return 0.01f;
+}
+
+template<>
+double eps<double>() {
+    return 1e-5;
+}
+
+template<>
+double eps<cfloat>() {
+    return 0.01f;
+}
+
+template<>
+double eps<cdouble>() {
+    return 1e-5;
+}
+
+template<typename T>
+double relEps(array in) {
+    typedef typename af::dtype_traits<T>::base_type InBaseType;
+    return std::numeric_limits<InBaseType>::epsilon()
+        * std::max(in.dims(0), in.dims(1)) * af::max<double>(in);
+}
+
+typedef ::testing::Types<float, cfloat, double, cdouble> TestTypes;
+TYPED_TEST_CASE(Pinverse, TestTypes);
+
+// Test Moore-Penrose conditions
+// See https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Definition
+
```

Comment:
Nit: Extra line

Reply:
Same reasoning as in api/c/pinverse.cpp but I just removed the extra line here and clarified that this comment just applies to the first 4 following tests